### PR TITLE
Access manager: Migrate user autocomplete UI to the wildcard combobox

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/41636

## Test plan
- Enable new invite users page for org (simple override context to `true` in [`routes.tsx`](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=vk%2Fmigrate-invite-users-picker&file=client%2Fweb%2Fsrc%2Forg%2Farea%2Froutes.tsx&start_row=23&start_col=47&end_row=23&end_col=70&editor=JetBrains&version=v2.0.1))
- Go to the org site admin page 
- Go to the members page
- Open invite member modal 
- Check that autocomplete works as expected (you can type user name and click invite, or you can use suggestions and pick user from these suggestions) 

## App preview:

- [Web](https://sg-web-vk-migrate-invite-users-picker.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-euvmfsrogi.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/wildcard/src/components/Combobox/Combobox.tsx [hunk 1]
```diff
@@ -145,5 +147,27 @@ export const ComboboxOptionGroup = forwardRef((props, ref) => {
...
+        return (
+            <li ref={ref} {...attributes}>
+                {typeof children === 'function' ? children(context) : children}
+            </li>
+        )
+    }
+
+    return (
+        <ReachComboboxOption ref={ref} {...attributes}>
+            {children}
```
Comment (valerybugakov) [inline client/wildcard/src/components/Combobox/Combobox.tsx:168]: Does it make sense to account for callable `children` as we do above?

```suggestion
            {typeof children === 'function' ? children(context) : children}
```

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 2]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
+                        {user.displayName && <span className={styles.userName}>{user.username}</span>}
+                    </div>
+                    {user.inOrg && <small className="text-muted">Already in this organization</small>}
+                </div>
+            </div>
+        </ComboboxOption>
     )
 }
+
+const EmptyResultsItem: FC<{ userNameOrEmail: string }> = ({ userNameOrEmail }) => (
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:136]:
```suggestion
interface EmptyResultsItemProps {
  userNameOrEmail: string
}

const EmptyResultsItem: FC<EmptyResultsItemProps> = ({ userNameOrEmail }) => (
```

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 3]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
-                                onKeyDown={onResultItemKeydown}
-                            />
-                        ))}
-                    {renderNoMatch && <EmptyResultsItem userNameOrEmail={userNameOrEmail} />}
-                </DropdownMenu>
-            </Dropdown>
-        </div>
+        <ComboboxOption
+            value={user.displayName ?? user.username}
+            data-testid="search-context-menu-item"
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:109]: I see that these `data-testid` attributes were here before the PR. Still, they are redundant, and it seems we can remove them without additional time investment to make clear that this component is not related to `search-context-menu` in any way.

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 4]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
+                            user={user}
+                        />
+                    </Tooltip>
+                </div>
+                <div className="d-flex flex-column">
+                    <div>
+                        <strong>{user.displayName || user.username}</strong>{' '}
+                        {user.displayName && <span className={styles.userName}>{user.username}</span>}
+                    </div>
+                    {user.inOrg && <small className="text-muted">Already in this organization</small>}
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:129]: This text is almost invisible for disabled hovered `UserResultItem`. What do you think about removing the `text-muted` class name for disabled hovered options?

Screenshot 2022-09-16 at 13 24 18

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 5]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
-        [firstResult]
-    )
+    const results = (data
+        ? data.autocompleteMembersSearch.map(usr => ({ ...usr })).sort(item => (item.inOrg ? 1 : -1))
+        : []) as IUserItem[]
 
-    const onSelectUser = useCallback(
-        (user: IUserItem) => {
-            eventLogger.log(
-                'InviteAutocompleteUserSelected',
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:179]: Do we want to preserve this event log?

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 6]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
-                                onSelectUser={onSelectUser}
-                                onKeyDown={onResultItemKeydown}
-                            />
-                        ))}
-                    {renderNoMatch && <EmptyResultsItem userNameOrEmail={userNameOrEmail} />}
-                </DropdownMenu>
-            </Dropdown>
-        </div>
+        <ComboboxOption
+            value={user.displayName ?? user.username}
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:108]: Users can often have the same display name, which results in confusing UI for focused options. What do you think about giving `user.username` a priority here?

Screenshot 2022-09-16 at 13 41 55

Screenshot 2022-09-16 at 13 42 18

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 7]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
+                        <UserAvatar
+                            size={24}
+                            className={classNames(styles.avatar, user.inOrg ? styles.avatarDisabled : undefined)}
+                            user={user}
+                        />
+                    </Tooltip>
+                </div>
+                <div className="d-flex flex-column">
+                    <div>
+                        <strong>{user.displayName || user.username}</strong>{' '}
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:126]: What do you think about improving contrast here for enabled hovered options? 

Screenshot 2022-09-16 at 13 45 25

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 8]
```diff
@@ -145,97 +48,104 @@ export const AutocompleteSearchUsers: React.FunctionComponent<
...
+                    {user.inOrg && <small className="text-muted">Already in this organization</small>}
+                </div>
+            </div>
+        </ComboboxOption>
     )
 }
+
+const EmptyResultsItem: FC<{ userNameOrEmail: string }> = ({ userNameOrEmail }) => (
+    <div
+        data-testid="search-context-menu-item"
```
Comment (valerybugakov) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:138]: Similar note: this component is not related to `search-context-menu`, meaning that this `data-testid` prop can be deleted.

File: client/wildcard/src/components/Combobox/Combobox.tsx [hunk 9]
```diff
@@ -145,5 +147,27 @@ export const ComboboxOptionGroup = forwardRef((props, ref) => {
...
+    disabled?: boolean
+}
+
+export const ComboboxOption = forwardRef((props, ref) => {
+    const { disabled, children, ...attributes } = props
+    const context = useComboboxOptionContext()
+
+    if (disabled) {
+        return (
+            <li ref={ref} {...attributes}>
```
Comment (valerybugakov) [inline client/wildcard/src/components/Combobox/Combobox.tsx:160]: At first, I tested this PR on the `sourcegraph` organization, and all suggestions were already in the org — all options were disabled, and it was hard to understand from the UI. For a minute, I thought that it was a bug. Would it be worth changing the cursor type for disabled options to `not-allowed` to make it more explicit?

**Review by valerybugakov — APPROVED** [review]
Functionality-wise, everything looks great! I noticed several UI/UX issues, which were probably here before this PR. Feel free to ignore them if fixes require a significant time investment.

File: client/wildcard/src/components/Combobox/Combobox.tsx (hunk 1 above)
Comment (vovakulikov) [reply client/wildcard/src/components/Combobox/Combobox.tsx:168]: @valerybugakov ReachCombobox will handle it for us. Internally It does the same check with typeof children as we do above.

File: client/web/src/org/members/SearchUserAutocomplete.tsx (hunk 3 above)
Comment (vovakulikov) [reply client/web/src/org/members/SearchUserAutocomplete.tsx:109]: Agree, I will remove this shortly. Thanks for spotting this.

File: client/web/src/org/members/SearchUserAutocomplete.tsx (hunk 4 above)
Comment (vovakulikov) [reply client/web/src/org/members/SearchUserAutocomplete.tsx:129]: This is something that was before this PR I believe. But I agree with the visual issue here. I will try to look at, thanks

File: client/web/src/org/members/SearchUserAutocomplete.tsx (hunk 5 above)
Comment (vovakulikov) [reply client/web/src/org/members/SearchUserAutocomplete.tsx:179]: My bad, I forgot to restore it. Thanks for catching this, I will bring this back.

File: client/wildcard/src/components/Combobox/Combobox.tsx (hunk 9 above)
Comment (vovakulikov) [reply client/wildcard/src/components/Combobox/Combobox.tsx:160]: Good idea. I will add disabled styles for it. Thanks

File: client/wildcard/src/components/Combobox/Combobox.tsx [hunk 10]
```diff
@@ -88,7 +90,7 @@ export const ComboboxPopover = forwardRef<HTMLDivElement, ComboboxPopoverProps>(
 
     // If we don't have registered input element we should not
     // render anything about combobox suggestions (popover content)
```
Comment (unclejustin) [inline client/wildcard/src/components/Combobox/Combobox.tsx:92]: This changes the logic a bit. Now `isOpen` is always true, which makes sense because of the early return here.

Could you update this comment to explain why we return `null` now if it's not expanded?

File: client/web/src/org/members/SearchUserAutocomplete.tsx [hunk 11]
```diff
@@ -1,118 +1,31 @@
...
+const MIN_SEARCH_LENGTH = 3
+const EMAIL_PATTERN = new RegExp(/^[\w!#$%&'*+./=?^`{|}~-]+@[A-Z_a-z]+?\.[A-Za-z]{2,3}$/)
 
 interface AutocompleteSearchUsersProps {
     disabled?: boolean
     onValueChanged: (value: string, isEmail: boolean) => void
     orgId: string
 }
 
-const UserResultItem: React.FunctionComponent<
```
Comment (unclejustin) [inline client/web/src/org/members/SearchUserAutocomplete.tsx:33]: Nitpick: Looks like this component got moved and edited in the same commit. Makes it harder to see the changes. WDYT about a quick 'refactoring' commit before making changes in future PRs? Separating simple refactors from actual changes would be super helpful to reviewers.

**Review by unclejustin — APPROVED** [review]
LGTM @vovakulikov 🎉 

Just one ask to update a code comment and a request to separate refactorings from actual code changes in the future.
